Adding howto to set up french fiber provider SFR/RED (#179)
opnsense/docs: source/manual/how-tos/sfr_red_fr_ftth.rst

SFR/RED France FTTH IPv4 & IPv6 & Phone
=======================================

**Original Author:** Philippe Gaultier

**Introduction**
-----------------

> This guide is for SFR/RED France FTTH using DHCP to connect

Author: Merge the two paragraphs in last commit

> The guide deals with internet connection and phone. Support for TV has not been tested

Author: Done in last commit

.. Note::

>     Before starting this guide, you should have the MAC adress of your SFR/RED Box. In the guide you should replace xx:xx:xx:xx:xx:xx by your SFR/RED Box MAC adress

Reviewer: ```suggestion
    Before starting this guide, you should have the MAC address of your SFR/RED Box. In this guide you should replace xx:xx:xx:xx:xx:xx with your SFR/RED Box MAC address
```

>     Before starting this guide, you should have the MAC adress of your SFR/RED Box. In the guide you should replace xx:xx:xx:xx:xx:xx by your SFR/RED Box MAC adress

Author: Fixed in last commit

**Getting ready to make the connection**
----------------------------------------

> SFR/RED requires that the WAN interface assignment should hence look something like this

Reviewer: ```suggestion
SFR/RED requires that the WAN interface assignment should look similar to this:
```

Author: Fixed

.. image:: images/SFRRED_assignations.png
	:width: 100%

* WAN interface has MAC xx:xx:xx:xx:xx:xx which is the original WAN MAC of the BOX (spoofed)
* LAN interface has MAC 00:11:22:33:44:55 which is the original MAC of the firewall
* DUID is 00:03:00:01:xx:xx:xx:xx:xx:xx it's derived from the original WAN MAC of the BOX (spoofed)

**Configuring the WAN Interface**
---------------------------------

Select :menuselection:`Interfaces --> [WAN]`

> In order to establish the IPv4 and IPv6 connection SFR/RED requires that the correct parameters are passed for the DHCP and DHCP6

Reviewer: ```suggestion
In order to establish the IPv4 and IPv6 connection, SFR/RED requires that the correct parameters are passed for the DHCPv4 and DHCPv6
```

> In order to establish the IPv4 and IPv6 connection SFR/RED requires that the correct parameters are passed for the DHCP and DHCP6

Author: Fixed in last commit

> requests respectively

Author: Added dots in last commit

> select options DHCP and DHCPv6 in general configuration

Reviewer 2: Even with the dots it looks weird:
```suggestion
Select options DHCP and DHCPv6 in general configuration:
```

Author: Updated in last commit

.. image:: images/SFRRED_WAN_configuration_1.png
	:width: 100%

**On the DHCP request it is a requirement to pass the following:**

.. image:: images/SFRRED_WAN_configuration_2.png
	:width: 100%

.. Note::

>     It is necessary to specify the following 'Send Options'

Reviewer: The docs use “these quotes” (or at least most of it, might have overlooked some).

Author: Done for the quotes - it was a copy/paste from another documentation :-)

* dhcp-class-identifier "neufbox_NB6VAC-FXC"

.. Note::
    It is necessary to specify the following 'Request Options'

> * subnet-mask, broadcast-address, time-offset, routers, domain-name, domain-name-servers, host-name, ntp-servers, nis-domain, root-path, merit-dump

Reviewer 2: in the source this looks strange for notes. Doesn't that need an indent?

Author: Added the options in Notes.

**On the DHCP6 request we need to use raw options**

.. image:: images/SFRRED_WAN_configuration_3.png
	:width: 100%

.. Note::
    It is necessary to specify the following 'Send Options'

* ia-pd 1, raw-option 16 00:00:a0:0c:00:40:6e:65:75:66:62:6f:78:5f:4e:42:36:56:41:43:2d:46:58:43

.. Note::
    It is necessary to specify the following 'Request Options'

* domain-name-servers, domain-name

> Finally set the Identity Association and Prefix interface as shown

Reviewer: You'll need to be a bit clearer here. What are you specifically referring to? An earlier screenshot or table, for example?

Author: Adding parameters to update in Note field

Click ‘Save’ and then ‘Apply’.

**Configuring the LAN Interface**
---------------------------------

Interfaces / Parameters
+++++++++++++++++++++++

> Select :menuselection:`Interfaces --> Parameters` and set your DUID

Reviewer: You need to end the sentences with a dot. Like the previous one, I won't spam this line note. ;-)

Author: Should be done for all sentences

.. image:: images/SFRRED_interfaces_parameters.png
	:width: 100%

.. Note::
    The DUID is based on the SFR/RED Box MAC address : 00:03:00:01:xx:xx:xx:xx:xx:xx

Click ‘Save’ and then ‘Apply’.

Interfaces / [LAN]
++++++++++++++++++

Select :menuselection:`Interfaces --> [LAN]` and set IPv4 to “Static IPv4” and IPv6 Configuration Type to
“Track Interface”.

.. image:: images/SFRRED_LAN_configuration_1.png
	:width: 100%

Finally, set the Track IPv6 Interface to WAN and set the IPv4 address to your chosen address.

.. image:: images/SFRRED_LAN_configuration_2.png
	:width: 100%

Click ‘Save’ and then ‘Apply’.

.. Note::

>     It is advisable at this point to reboot the system. This will allow you to retrieve IPv4 which will be used in next part

Reviewer: ```suggestion
    It is advisable at this point to reboot the system. This will allow you to retrieve an IPv4 address which will be used in next part
```
(assuming that this indeed refers to an IPv4 address and not something else, of course)

Author: Fixed

**Configuring NGINX to provision the SFR/RED BOX**
--------------------------------------------------

In order to set up the phone, as the SIP parameters (user/password) are not public, we will add the SFR/RED box in our LAN.
This will allow us to plug our regular phone in the SFR/RED box.

.. Note::
    This how-to does not cover installation of NGINX nor the use of SSH / shell commands.

> First login into you OPNSense firewall and create a folder **/srv/sfrredbox**. We will add in this folder the scripts used to spoof the SFR/RED Box requests.

Reviewer: ```suggestion
First SSH into your OPNSense firewall and create a folder **/srv/sfrredbox**. In this folder, we will add the scripts used to spoof the SFR/RED Box requests.
```

Author: Fixed

.. code-block:: php

    $currentFirewall = 'firewall.localdomain.intra';
    // can probably be replaced with
    // $currentFirewall = exec('hostname');
    if (isset($_GET['ip_dhcp'])) {
        // adjust re0 to your WAN interface
        $_GET['ip_dhcp'] = exec('ifconfig re0 | grep \'inet \' | cut -d\' \' -f2');
        // if the ifconfig command does not work, set the external IP manually
        // $_GET['ip_dhcp'] = 'your.external.ip.address';
    }
    $_SERVER['DOCUMENT_URI'] = str_replace('/index.php', '', $_SERVER['DOCUMENT_URI']);
    $parameters = http_build_query($_GET);
    $url = $_SERVER['REQUEST_SCHEME'].'://'.$_SERVER['HTTP_HOST'].'/'.trim($_SERVER['DOCUMENT_URI'], '/?') .'?'.$parameters;
    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $url);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_HEADERFUNCTION, 'readHeaderLine');
    $data = curl_exec($ch);
    $data = preg_replace('/<proxy([^>]+)>([^<]+)<\/proxy>/', '<proxy$1>'.$currentFirewall.'</proxy>', $data);
    curl_close($ch);
    header('Content-Length: '.strlen($data));
    header('Content-Type: application/xml');
    echo $data;

.. Warning::

>     Code cannot be copy / paste, you have to adjust the parameters and make it consistant with your own parameters

Reviewer: ```suggestion
    Code cannot be copied and pasted as-is, you will have to adjust the parameters and make it consistent with your own settings.
```

Author: fixed

Services / Nginx / Configuration
++++++++++++++++++++++++++++++++

Select :menuselection:`Services --> Nginx --> Configuration`

Activate NGINX

.. image:: images/SFRRED_services_nginx_configuration_1.png
	:width: 100%

Services / Nginx / Configuration / HTTP(s)
++++++++++++++++++++++++++++++++++++++++++

Select :menuselection:`Services --> Nginx --> Configuration --> HTTP(s)`

Create a new config

> .. image:: images/SFRRED_services_nginx_configuration_2.png

Reviewer 2: you should select the match type in the screen shot in the next line. Abusing fields may lead to unwanted side effects in future releases.

BTW: matching on "/" should do as well if you don't care about the dots.

Author: Well, I tried several configurations and this one is working as expected.
About the fields, this is a full screenshot because I thought it would be easier for the reader to see the whole page and know where he can leave fileds 'blank'

.. highlights::
    host -t SRV _sip._udp.residential.p-cscf.sfr.net

.. Note::
    the host request result gives available SIP servers with the port to use (in my case 5062)

.. highlights::
    _sip._udp.residential.p-cscf.sfr.net has SRV record 10 0 5062 mitry.p-cscf.sfr.net.
    _sip._udp.residential.p-cscf.sfr.net has SRV record 10 0 5062 corbas.p-cscf.sfr.net.
    _sip._udp.residential.p-cscf.sfr.net has SRV record 10 0 5062 trappes.p-cscf.sfr.net.

**Configuring NAT to redirect SFR/RED BOX calls to NGINX**
----------------------------------------------------------

To allow correct port forwarding, we will configure OPNSense to affect a **static** IP to the SFR/RED Box and we will create an alias for it.

Services / DHCPv4 / [LAN]
+++++++++++++++++++++++++

Select :menuselection:`Services --> DHCPv4 --> [LAN]`

Click on `[+]` to add a static mapping

.. image:: images/SFRRED_services_dhcp_lan.png
	:width: 100%

Firewall / NAT / Port Forward
+++++++++++++++++++++++++++++

Select :menuselection:`Firewall --> NAT --> Port Forward`

Add a new forwarding rule

.. image:: images/SFRRED_lan_port_forwarding.png
	:width: 100%

.. Note::

>     Right now, everything should be ready. Restart the firewall, once ready plug the SFR/RED Box on you LAN and start it.

Reviewer: ```suggestion
    Right now, everything should be ready. Restart the firewall, once ready plug the SFR/RED Box on your LAN and start it.
```

Author: Fixed
